Mpi/fix global pointer communications

roigcarlo opened the pull request:
Solving some errors...

@ddiezrod you will need this to run the new codes for the non_cluster_meeting process and the cluster algorithms without error.

philbucher commented on kratos/utilities/global_pointer_utilities.h:
```diff
@@ -363,16 +362,20 @@ class GlobalPointerUtilities
 
     //particularizing to the case of nodes
-    static bool IteratorIsLocal(ModelPart::NodesContainerType::iterator& it, const int CurrentRank)
+    static bool IteratorIsLocal(Node<3>& node, const int CurrentRank)
```
```suggestion
    static bool ObjectIsLocal(const Node<3>& rNode, const int CurrentRank)
```

philbucher commented on kratos/utilities/global_pointer_utilities.h:
```diff
@@ -363,16 +362,20 @@ class GlobalPointerUtilities
+    static bool IteratorIsLocal(Node<3>& node, const int CurrentRank)
     {
-        if(it->FastGetSolutionStepValue(PARTITION_INDEX) == CurrentRank)
+        if(node.FastGetSolutionStepValue(PARTITION_INDEX) == CurrentRank)
```
```suggestion
        return node.FastGetSolutionStepValue(PARTITION_INDEX) == CurrentRank;
```

philbucher commented on kratos/utilities/global_pointer_utilities.h:
```diff
@@ -363,16 +362,20 @@ class GlobalPointerUtilities
     ///@{
 
-    template< class TIteratorType >
-    static bool IteratorIsLocal(TIteratorType& it, const int CurrentRank)
```
why do you change this?
Looks correct to me

roigcarlo replied:
Is correct. In fact this is the problem.
The function is correct and is so generic that everything can enter there. The problem comes when you have a let's say a node iterator

You would expect that to enter in: 

`static bool IteratorIsLocal(ModelPart::NodesContainerType::iterator& it, const int CurrentRank)`

which checks if the node belongs to the local partition or nor, but in fact you may have:

`const ModelPart::NodesContainerType::iterator &`
`ModelPart::NodesContainerType::const_iterator &`
`const ModelPart::NodesContainerType::const iterator &`

And all of them will by default resolve to use the generic variant, which will casue all the nodes to be treated as if they were from the local partition, whereas they are or not.

By explicitly typifying the function, at least, we can cause a compilation error if its not possible to call it, which will point to the error, instead of returining `True` by default.

philbucher replied:
ok makes sense :)

RiccardoRossi replied:
nasty one (i'll take the blame for this)

philbucher reviewed (changes requested):
minor cleanup required

roigcarlo committed:
Compilation fixes

RiccardoRossi commented on kratos/utilities/global_pointer_utilities.h:
```diff
@@ -97,8 +97,7 @@ class GlobalPointerUtilities
             {
-                //if(container[id].FastGetSolutionStepValue(PARTITION_INDEX) == current_rank)
-                if(IteratorIsLocal(it, current_rank))
+                if(!rDataCommunicator.IsDistributed() || ObjectIsLocal(*it, current_rank))
```
can t we do the check for is_distributed only.once instead of per every iterstor?

roigcarlo replied:
mm you are right :S

RiccardoRossi reviewed (commented):
the isdisgri uted check should be out of the loop

roigcarlo committed:
Preventing serializer error

Prevents error serializing the global_pointer to a string and uses the
default global_pointer save/load functions (which already take into account the SHALLOW_COPY flag)

roigcarlo committed:
Fixing error making non-const non-const-iterators (its not a typo) return always true. Funny one

roigcarlo committed:
Typo

philbucher commented:
It seems like there is no test for this
can you please add one?

roigcarlo committed:
Suggestions

roigcarlo committed:
adding consts

RiccardoRossi commented on kratos/containers/global_pointers_vector.h:
```diff
@@ -455,61 +455,23 @@ class GlobalPointersVector
-            }
+        rSerializer.load("Size", size);
+            
+        for(std::size_t i = 0; i < size; i++) {
```
here we may want to do a reserve of this

philbucher commented:
@RiccardoRossi please approve if you are ok, then I will also approve

roigcarlo commented:
All changes adressed. Ping @RiccardoRossi

roigcarlo committed:
Adding test to check that remote GP are fetched when there are local copies with PARTITION_INDEX != current_rank

roigcarlo committed:
Optimizations

roigcarlo committed:
Warning
